File: android-core/src/test/java/com/mparticle/networking/DomainMappingTest.java
package com.mparticle.networking;

import com.mparticle.testutils.RandomUtils;

import org.junit.Test;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Random;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;

public class DomainMappingTest {
    Random random = new Random();
    RandomUtils randomUtils = new RandomUtils();

    @Test
    public void nullCertificatesNotAdded() {
        List<Certificate> certificates = new ArrayList<Certificate>();
        certificates.add(null);
        DomainMapping domainMapping = DomainMapping.eventsMapping("test")
                .setCertificates(certificates)
                .build();
        assertEquals(0, domainMapping.getCertificates().size());
    }

    @Test
    public void malformedCertificateNotAdded() {
        DomainMapping domainMapping = DomainMapping.eventsMapping("test")
                .addCertificate("", "")
                .build();
        assertEquals(0, domainMapping.getCertificates().size());
    }

    @Test
    public void addCertificatTest() {
        Certificate certificate = Certificate.with("alias", "certificate");
        DomainMapping domainMapping = DomainMapping.eventsMapping("test")
                .addCertificate(certificate.getAlias(), certificate.getCertificate())
                .build();

        assertEquals(1, domainMapping.getCertificates().size());
        assertEquals(certificate.getAlias(), domainMapping.getCertificates().get(0).getAlias());
        assertEquals(certificate.getCertificate(), domainMapping.getCertificates().get(0).getCertificate());
    }

    @Test
    public void addMultipleCertificatesTest() {
        Map<String, Certificate> certificates = new HashMap<String, Certificate>();

        for (int i = 0; i < random.nextInt() % 10; i++) {
            String alias = randomUtils.getAlphaNumericString(10);
            certificates.put(alias, Certificate.with(alias, randomUtils.getAlphaNumericString(124)));
        }

        DomainMapping.Builder builder = DomainMapping.eventsMapping("test");
        for (Certificate certificate: certificates.values()) {
            builder.addCertificate(certificate);
        }

        DomainMapping domainMapping = builder.build();

        assertContainsCertificates(certificates, domainMapping);

        domainMapping = DomainMapping.eventsMapping("test")
                .setCertificates(new ArrayList<Certificate>(certificates.values()))
                .build();

        assertContainsCertificates(certificates, domainMapping);

        builder = DomainMapping.eventsMapping("test");
        for (Certificate certificate: certificates.values()) {
            builder.addCertificate(certificate.getAlias(), certificate.getCertificate());
        }
        domainMapping = builder.build();

        assertContainsCertificates(certificates, domainMapping);
    }

    private void assertContainsCertificates(Map<String, Certificate> certificates, DomainMapping domainMapping) {
        assertEquals(certificates.values().size(), domainMapping.getCertificates().size());
        for(Certificate certificate: domainMapping.getCertificates()) {
            Certificate match = certificates.get(certificate.getAlias());
            assertNotNull(match);
            assertEquals(match.getAlias(), certificate.getAlias());
            assertEquals(match.getCertificate(), certificate.getCertificate());
        }
    }
}

File: android-core/src/test/kotlin/com/mparticle/networking/DomainMappingTest.kt
package com.mparticle.networking

import com.mparticle.testutils.RandomUtils
import org.junit.Assert
import org.junit.Test
import java.util.Random

class DomainMappingTest {
    private var random = Random()
    var randomUtils = RandomUtils()

    @Test
    fun nullCertificatesNotAdded() {
        val certificates: MutableList<Certificate?> = ArrayList()
        certificates.add(null)
        val domainMapping = DomainMapping.eventsMapping("test")
            .setCertificates(certificates)
            .build()
        Assert.assertEquals(0, domainMapping.certificates.size.toLong())
    }

    @Test
    fun malformedCertificateNotAdded() {
        val domainMapping = DomainMapping.eventsMapping("test")
            .addCertificate("", "")
            .build()
        Assert.assertEquals(0, domainMapping.certificates.size.toLong())
    }

    @Test
    fun addCertificateTest() {
        val certificate = Certificate.with("alias", "certificate")
        val domainMapping = DomainMapping.eventsMapping("test")
            .addCertificate(certificate!!.alias, certificate.certificate)
            .build()
        Assert.assertEquals(1, domainMapping.certificates.size.toLong())
        Assert.assertEquals(certificate.alias, domainMapping.certificates[0].alias)
        Assert.assertEquals(certificate.certificate, domainMapping.certificates[0].certificate)
    }

    @Test
    fun addMultipleCertificatesTest() {
        val certificates = HashMap<String, Certificate?>()
        for (i in 0 until random.nextInt() % 10) {
            val alias = randomUtils.getAlphaNumericString(10)
            certificates[alias] = Certificate.with(alias, randomUtils.getAlphaNumericString(124))
        }
        var builder = DomainMapping.eventsMapping("test")
        for (certificate in certificates.values) {
            if (certificate != null) {
                builder.addCertificate(certificate)
            }
        }
        var domainMapping = builder.build()
        assertContainsCertificates(certificates, domainMapping)
        domainMapping = DomainMapping.eventsMapping("test")
            .setCertificates(ArrayList(certificates.values))
            .build()
        assertContainsCertificates(certificates, domainMapping)
        builder = DomainMapping.eventsMapping("test")
        for (certificate in certificates.values) {
            certificate?.alias?.let { builder.addCertificate(it, certificate.certificate) }
        }
        domainMapping = builder.build()
        assertContainsCertificates(certificates, domainMapping)
    }

    private fun assertContainsCertificates(
        certificates: Map<String, Certificate?>,
        domainMapping: DomainMapping
    ) {
        Assert.assertEquals(
            certificates.values.size.toLong(),
            domainMapping.certificates.size.toLong()
        )
        for (certificate in domainMapping.certificates) {
            val match = certificates[certificate.alias]
            Assert.assertNotNull(match)
            Assert.assertEquals(match?.alias, certificate.alias)
            Assert.assertEquals(match?.certificate, certificate.certificate)
        }
    }
}
